File: ycmd/completers/javascript/tests/test_utils.py
#
# Copyright (C) 2015 ycmd contributors
#
# This file is part of YouCompleteMe.
#
# YouCompleteMe is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# YouCompleteMe is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.

import contextlib2, os, time
from ycmd.tests.test_utils import BuildRequest

TEST_DATA_DIR = os.path.join( os.path.dirname( __file__ ), 'testdata' )


@contextlib2.contextmanager
def with_cwd( wd ):
  # Sadly, the Tern.js server requires that its cwd is within the "project"
  # directory, so we have to set the working directory to our testdata directory
  # for each test. The following context manager just temporarily changes the
  # working directory to the supplied path.
  #
  # contextlib2 is used because it works in python 2.6 (as a decorator)
  prev_wd = os.getcwd()
  os.chdir( wd )
  try:
    yield
  finally:
    os.chdir( prev_wd )


def PathToTestFile( *args ):
  return os.path.abspath( os.path.join( TEST_DATA_DIR, *args ) )


def WaitForTernServerReady( app ):
  app.post_json( '/run_completer_command', BuildRequest(
    command_arguments = [ 'StartServer' ],
    completer_target = 'filetype_default',
    filetype = 'javascript',
    filepath = '/foo.js',
    contents = '',
    line_num = '1'
  ) )

  retries = 100
  while retries > 0:
    result = app.get( '/ready', { 'subserver': 'javascript' } ).json
    if result:
      return

    time.sleep( 0.2 )
    retries = retries - 1

  raise RuntimeError( 'Timeout waiting for Tern.js server to be ready' )

File: cpp/ycm/tests/TestUtils.cpp
// Copyright (C) 2011-2018 ycmd contributors
//
// This file is part of ycmd.
//
// ycmd is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// ycmd is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with ycmd.  If not, see <http://www.gnu.org/licenses/>.

#include "TestUtils.h"

#include <whereami.c>

namespace boost {

namespace filesystem {

void PrintTo( const fs::path &path, std::ostream *os ) {
  *os << path;
}

} // namespace filesystem

} // namespace boost

namespace YouCompleteMe {

std::ostream& operator<<( std::ostream& os, const CodePointTuple &code_point ) {
  os << "{ " << PrintToString( code_point.normal_ ) << ", "
             << PrintToString( code_point.folded_case_ ) << ", "
             << PrintToString( code_point.swapped_case_ ) << ", "
             << PrintToString( code_point.is_letter_ ) << ", "
             << PrintToString( code_point.is_punctuation_ ) << ", "
             << PrintToString( code_point.is_uppercase_ ) << ", "
             << PrintToString( code_point.break_property_ ) << " }";
  return os;
}

std::ostream& operator<<( std::ostream& os, const CodePoint &code_point ) {
  os << CodePointTuple( code_point );
  return os;
}


std::ostream& operator<<( std::ostream& os, const CodePoint *code_point ) {
  os << "*" << *code_point;
  return os;
}


std::ostream& operator<<( std::ostream& os, const CharacterTuple &character ) {
  os << "{ " << PrintToString( character.normal_ ) << ", "
             << PrintToString( character.base_ ) << ", "
             << PrintToString( character.folded_case_ ) << ", "
             << PrintToString( character.swapped_case_ ) << ", "
             << PrintToString( character.is_base_ ) << ", "
             << PrintToString( character.is_letter_ ) << ", "
             << PrintToString( character.is_punctuation_ ) << ", "
             << PrintToString( character.is_uppercase_ ) << " }";
  return os;
}


std::ostream& operator<<( std::ostream& os, const Character &character ) {
  os << PrintToString( CharacterTuple( character ) );
  return os;
}


std::ostream& operator<<( std::ostream& os, const Character *character ) {
  os << "*" << *character;
  return os;
}


std::ostream& operator<<( std::ostream& os, const WordTuple &word ) {
  os << "{ " << PrintToString( word.text_ ) << ", { ";
  const std::vector< const char* > &characters( word.characters_ );
  auto character_pos = characters.begin();
  if ( character_pos != characters.end() ) {
    os << PrintToString( *character_pos );
    ++character_pos;
    for ( ; character_pos != characters.end() ; ++character_pos ) {
      os << ", " << PrintToString( *character_pos );
    }
    os << " }";
  }
  return os;
}


std::ostream& operator<<( std::ostream& os, const fs::path *path ) {
  os << *path;
  return os;
}


fs::path PathToTestFile( const std::string &filepath ) {
  int dirname_length;
  int exec_length = wai_getExecutablePath( NULL, 0, NULL );
  std::unique_ptr< char[] > executable( new char [ exec_length ] );
  wai_getExecutablePath( executable.get(), exec_length, &dirname_length );
  executable[ dirname_length ] = '\0';
  fs::path path_to_testdata = fs::path( executable.get() ) / "testdata";
  return path_to_testdata / fs::path( filepath );
}

} // namespace YouCompleteMe
